File: backend/schemas/story_package.py
"""
Story Package Schema — The shared contract (specs/story-package.md).

Every agent reads and writes these structured objects.
Free-form agent prose must never pass directly into child mode.
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field, field_validator


# ── Lifecycle ──────────────────────────────────────────────────────────────

class StoryStatus(str, Enum):
    CAPTURED = "captured"
    INTERPRETING = "interpreting"
    NEEDS_CLARIFICATION = "needs_clarification"  # F3 — paused for one parent answer
    PLANNING = "planning"
    WRITING = "writing"
    VALIDATING = "validating"
    AWAITING_PARENT = "awaiting_parent"
    APPROVED = "approved"
    IN_SESSION = "in_session"
    COMPLETED = "completed"


class ValidationStatus(str, Enum):
    PENDING = "pending"
    PASSED = "passed"
    REVISE = "revise"
    BLOCKED = "blocked"


class FamilyVoiceMode(str, Enum):
    CONFIDENT = "confident"
    LEARNING = "learning"
    PRERECORDED = "prerecorded"


class InteractionType(str, Enum):
    CHOICE = "choice"
    SPEAK = "speak"


class InputType(str, Enum):
    TEXT = "text"
    VOICE = "voice"
    PHOTO = "photo"


class ConfidenceLevel(str, Enum):
    BEGINNING = "beginning"
    EMERGING = "emerging"
    GROWING = "growing"
    CONVERSATIONAL = "conversational"


# ── Sub-schemas ────────────────────────────────────────────────────────────

class MomentFact(BaseModel):
    text: str
    confidence: float = Field(ge=0.0, le=1.0)


class Clarification(BaseModel):
    """F3 — one follow-up question when the moment is too vague."""
    needed: bool = False
    question: str = ""
    answer: str = ""


class LanguageInfo(BaseModel):
    locale: str  # ta-SG, zh-SG, ms-SG
    pack_version: str = "1.0.0"


class LearningPlan(BaseModel):
    speaking_goal: str
    target_words: list[str] = Field(min_length=3, max_length=5)
    target_phrase: str
    level: ConfidenceLevel
    expected_intents: list[str] = []


# Child mode must never present more than 3 tappable choices at once
# (specs/story-package.md · bounded interactions).
MAX_CHOICES = 3


class SceneInteraction(BaseModel):
    type: InteractionType
    options: list[str] = []
    expected_intent: str = ""

    @field_validator("options")
    @classmethod
    def _cap_options(cls, v: list[str]) -> list[str]:
        # Hard cap — protects the child UI even if an agent/LLM over-produces.
        return v[:MAX_CHOICES]


class StoryScene(BaseModel):
    index: int
    # A chapter name the way a storybook has one — "The Slipper in the Bushes",
    # not "Scene 2". Shown above the page in child mode.
    title: str = ""
    title_target_lang: str = ""
    # One picture for this beat of the story, chosen by the writer so it
    # matches what actually happens here.
    emoji: str = ""
    narration: str = ""
    narration_target_lang: str = ""
    visual_id: str = ""
    interaction: SceneInteraction = SceneInteraction(type=InteractionType.CHOICE)


class RoomMission(BaseModel):
    instruction: str = ""
    instruction_target_lang: str = ""
    safety_validated: bool = False


class FamilyHandoff(BaseModel):
    mode: FamilyVoiceMode = FamilyVoiceMode.CONFIDENT
    prompt: str = ""
    prompt_target_lang: str = ""
    response_suggestion: str = ""
    parent_support: Optional[dict] = None


class EndingPrompt(BaseModel):
    text: str = ""


class VocabWord(BaseModel):
    """One "word to learn" from the story — spoken by Mina, repeated by the
    child. Native script is what the child sees; romanised helps the parent."""
    word: str  # English
    word_target_lang: str = ""
    romanised: str = ""


class Story(BaseModel):
    title: str = ""
    title_target_lang: str = ""
    opening_choices: list[str] = Field(default_factory=list, max_length=MAX_CHOICES)
    scenes: list[StoryScene] = Field(default_factory=list)
    room_mission: RoomMission = RoomMission()
    family_handoff: FamilyHandoff = FamilyHandoff()
    ending_prompt: EndingPrompt = EndingPrompt()
    # Words to learn — the learning plan's target words in the home language
    vocabulary: list[VocabWord] = Field(default_factory=list)
    # The sing-song line that comes back scene after scene, word for word, so
    # the child can chant it along with Mina — what makes four scenes feel
    # like one story instead of four sentences.
    refrain: str = ""
    refrain_target_lang: str = ""
    # The one hero, named once. Translated a single time and then spelled that
    # same way in every scene — a child who hears three names hears three
    # different stories.
    hero_name: str = ""
    hero_name_target_lang: str = ""


class NarrationSegment(BaseModel):
    scene_index: int
    tts_provider: str = ""
    audio_url: str = ""
    text: str = ""
    text_target_lang: str = ""


class SceneAsset(BaseModel):
    visual_id: str = ""
    asset_url: str = ""


class MediaAsset(BaseModel):
    """F4 — one pre-generated audio asset in the media manifest.
    url is relative to the API prefix (e.g. media/{pkg}/{id}.wav);
    empty url == parent-read fallback (text + romanization)."""
    id: str
    kind: str  # scene | mission | handoff | feedback | vocab
    scene_index: int = -1
    url: str = ""
    duration_ms: int = 0
    tts_provider: str = "text_only"
    text: str = ""
    text_target_lang: str = ""


class Media(BaseModel):
    narration_segments: list[NarrationSegment] = []
    scene_assets: list[SceneAsset] = []
    # F4 — TTS pre-generation on approval
    manifest: list[MediaAsset] = []
    manifest_ready: bool = False


class FamilyVoiceConfig(BaseModel):
    mode: FamilyVoiceMode = FamilyVoiceMode.CONFIDENT
    speaker_label: str = ""
    is_remote: bool = False


class Validation(BaseModel):
    language: ValidationStatus = ValidationStatus.PENDING
    safety: ValidationStatus = ValidationStatus.PENDING
    parent_approved_at: Optional[datetime] = None


class Provenance(BaseModel):
    agent_spec_versions: dict[str, str] = {}
    model_versions: dict[str, str] = {}


class SessionOutcome(BaseModel):
    """Growth Coach output persisted on session completion (F11).
    Non-judgmental: no scores, no proficiency labels — just a forward-looking
    seed for the next moment plus an encouraging note for the parent."""
    completed_at: Optional[datetime] = None
    next_moment_suggestion: str = ""
    encouragement: str = ""
    coach_source: str = ""  # qwen-max | fallback


# ── The Story Package ─────────────────────────────────────────────────────

class StoryPackage(BaseModel):
    """The central shared contract — specs/story-package.md."""
    id: str
    status: StoryStatus = StoryStatus.CAPTURED
    child_profile_id: str = ""
    moment_id: str = ""
    language: LanguageInfo = LanguageInfo(locale="ta-SG")

    # Raw parent description of the moment (input to Moment Lens)
    moment_text: str = ""
    moment_facts: list[MomentFact] = []
    # F3 — needs_clarification pause state
    clarification: Clarification = Clarification()
    learning_plan: Optional[LearningPlan] = None
    story: Story = Story()
    media: Media = Media()
    family_voice: FamilyVoiceConfig = FamilyVoiceConfig()
    validation: Validation = Validation()
    provenance: Provenance = Provenance()
    # F11 — Growth Coach session record + next-moment seed
    session_outcome: Optional[SessionOutcome] = None

    # Parent review & edit (F2): single-component regenerations, hard cap 5
    regeneration_count: int = 0

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


# ── Child Profile ──────────────────────────────────────────────────────────

class ChildProfile(BaseModel):
    id: str
    adult_id: str
    alias: str
    age_band: str  # "4-5", "6-7", "7-8"
    target_locale: str = "ta-SG"
    home_language: str = "en"  # language spoken at home (en/ta/zh/ms)
    photo_url: Optional[str] = None  # API path served by the backend
    understanding_level: ConfidenceLevel = ConfidenceLevel.EMERGING
    speaking_level: ConfidenceLevel = ConfidenceLevel.EMERGING
    interests: list[str] = []
    created_at: datetime = Field(default_factory=datetime.utcnow)
    deleted_at: Optional[datetime] = None


# ── Family Speaker ─────────────────────────────────────────────────────────

class FamilySpeaker(BaseModel):
    id: str
    child_profile_id: str
    relationship_label: str = ""  # "Amma", "Appa", "Thatha" — not "grandparent"
    mode: FamilyVoiceMode = FamilyVoiceMode.CONFIDENT
    language_confidence: str = ""
    is_remote: bool = False


# ── Moment ─────────────────────────────────────────────────────────────────

class Moment(BaseModel):
    id: str
    child_profile_id: str
    input_type: InputType
    parent_text: str = ""
    transcript: str = ""
    source_media_url: str = ""
    source_media_retention: str = "discard_after_generation"
    structured_facts: list[MomentFact] = []
    status: str = "captured"
    created_at: datetime = Field(default_factory=datetime.utcnow)


# ── Story Session ──────────────────────────────────────────────────────────

class StorySession(BaseModel):
    id: str
    story_package_id: str
    started_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: Optional[datetime] = None
    last_scene: int = 0
    speech_turn_completed: bool = False
    mission_completed: bool = False
    handoff_completed: bool = False
    duration_seconds: int = 0
    fallback_events: int = 0


# ── Saved Memory (F10) ───────────────────────────────────────────────────────

class SavedMemory(BaseModel):
    """A family memory saved with explicit consent (F10, AC-07).
    Deleting it removes this row AND the package's media files."""
    id: str
    session_id: str
    story_package_id: str
    child_profile_id: str = ""
    title: str = ""
    target_phrase: str = ""
    note: str = ""
    created_at: datetime = Field(default_factory=datetime.utcnow)


# ── Analytics Events ───────────────────────────────────────────────────────

class AnalyticsEvent(BaseModel):
    event_type: str
    child_profile_id: str = ""
    story_package_id: str = ""
    session_id: str = ""
    metadata: dict = {}
    created_at: datetime = Field(default_factory=datetime.utcnow)

File: app/lib/models/story_package.dart
/// TaleLah data models matching backend Story Package contract.

class ChildProfile {
  final String id;
  final String alias;
  final String ageBand;
  final String targetLocale;
  final String understandingLevel;
  final String speakingLevel;
  final List<String> interests;
  final String homeLanguage;
  final String? photoUrl;

  ChildProfile({
    required this.id,
    required this.alias,
    required this.ageBand,
    this.targetLocale = 'ta-SG',
    this.understandingLevel = 'emerging',
    this.speakingLevel = 'emerging',
    this.interests = const [],
    this.homeLanguage = 'en',
    this.photoUrl,
  });

  factory ChildProfile.fromJson(Map<String, dynamic> json) => ChildProfile(
        id: json['id'] as String,
        alias: json['alias'] as String,
        ageBand: json['age_band'] as String,
        targetLocale: json['target_locale'] as String? ?? 'ta-SG',
        understandingLevel: json['understanding_level'] as String? ?? 'emerging',
        speakingLevel: json['speaking_level'] as String? ?? 'emerging',
        interests: (json['interests'] as List<dynamic>?)
                ?.map((e) => e as String)
                .toList() ??
            [],
        homeLanguage: json['home_language'] as String? ?? 'en',
        photoUrl: json['photo_url'] as String?,
      );
}

class Moment {
  final String id;
  final String childProfileId;
  final String parentText;
  final String status;
  final String createdAt;

  Moment({
    required this.id,
    required this.childProfileId,
    required this.parentText,
    this.status = 'captured',
    this.createdAt = '',
  });

  factory Moment.fromJson(Map<String, dynamic> json) => Moment(
        id: json['id'] as String,
        childProfileId: json['child_profile_id'] as String,
        parentText: json['parent_text'] as String? ?? '',
        status: json['status'] as String? ?? 'captured',
        createdAt: json['created_at'] as String? ?? '',
      );
}

class StoryScene {
  final int index;
  final String title;
  final String narrationEnglish;
  final String narrationTargetLang;
  final String? audioUrl;
  final String interactionType;
  final String? promptText;
  final List<String> expectedResponses;

  StoryScene({
    required this.index,
    required this.title,
    this.narrationEnglish = '',
    this.narrationTargetLang = '',
    this.audioUrl,
    this.interactionType = 'listen',
    this.promptText,
    this.expectedResponses = const [],
  });

  factory StoryScene.fromJson(Map<String, dynamic> json) => StoryScene(
        index: json['index'] as int? ?? 0,
        title: json['title'] as String? ?? '',
        narrationEnglish: json['narration_english'] as String? ?? '',
        narrationTargetLang: json['narration_target_lang'] as String? ?? '',
        audioUrl: json['audio_data_uri'] as String?,
        interactionType: json['interaction_type'] as String? ?? 'listen',
        promptText: json['prompt_text'] as String?,
        expectedResponses: (json['expected_responses'] as List<dynamic>?)
                ?.map((e) => e as String)
                .toList() ??
            [],
      );
}

class StoryPackageSummary {
  final String id;
  final String status;
  final String childProfileId;
  final String languageLocale;
  final String title;
  final String speakingGoal;
  final String targetPhrase;
  final int sceneCount;
  final bool hasMission;
  final bool hasHandoff;
  final String validationLanguage;
  final String validationSafety;
  final String createdAt;
  /// Cover illustration URL (first scene's Wanx illustration).
  final String coverIllustrationUrl;

  StoryPackageSummary({
    required this.id,
    required this.status,
    required this.childProfileId,
    this.languageLocale = 'ta-SG',
    this.title = '',
    this.speakingGoal = '',
    this.targetPhrase = '',
    this.sceneCount = 0,
    this.hasMission = false,
    this.hasHandoff = false,
    this.validationLanguage = 'pending',
    this.validationSafety = 'pending',
    this.createdAt = '',
    this.coverIllustrationUrl = '',
  });

  factory StoryPackageSummary.fromJson(Map<String, dynamic> json) =>
      StoryPackageSummary(
        id: json['id'] as String,
        status: json['status'] as String,
        childProfileId: json['child_profile_id'] as String,
        languageLocale: json['language_locale'] as String? ?? 'ta-SG',
        title: json['title'] as String? ?? '',
        speakingGoal: json['speaking_goal'] as String? ?? '',
        targetPhrase: json['target_phrase'] as String? ?? '',
        sceneCount: json['scene_count'] as int? ?? 0,
        hasMission: json['has_mission'] as bool? ?? false,
        hasHandoff: json['has_handoff'] as bool? ?? false,
        validationLanguage: json['validation_language'] as String? ?? 'pending',
        validationSafety: json['validation_safety'] as String? ?? 'pending',
        createdAt: json['created_at'] as String? ?? '',
      );

  String get localeLabel {
    switch (languageLocale) {
      case 'ta-SG':
        return 'Tamil';
      case 'zh-SG':
        return 'Chinese';
      case 'ms-SG':
        return 'Malay';
      default:
        return languageLocale;
    }
  }
}

/// F4 — one pre-generated audio asset from the media manifest.
/// Empty [url] means parent-read fallback (adult reads the text aloud).
class MediaAsset {
  final String id;
  final String kind; // scene | mission | handoff
  final int sceneIndex;
  final String url;
  final int durationMs;
  final String ttsProvider;
  final String text;
  final String textTargetLang;

  MediaAsset({
    required this.id,
    required this.kind,
    this.sceneIndex = -1,
    this.url = '',
    this.durationMs = 0,
    this.ttsProvider = 'text_only',
    this.text = '',
    this.textTargetLang = '',
  });

  bool get hasAudio => url.isNotEmpty;

  factory MediaAsset.fromJson(Map<String, dynamic> json) => MediaAsset(
        id: json['id'] as String? ?? '',
        kind: json['kind'] as String? ?? '',
        sceneIndex: json['scene_index'] as int? ?? -1,
        url: json['url'] as String? ?? '',
        durationMs: json['duration_ms'] as int? ?? 0,
        ttsProvider: json['tts_provider'] as String? ?? 'text_only',
        text: json['text'] as String? ?? '',
        textTargetLang: json['text_target_lang'] as String? ?? '',
      );
}

/// F4 — a real approved scene for child-mode playback.
class ApprovedScene {
  final int index;
  /// Chapter name the way a storybook has one — the writer chose it for this
  /// beat, so it is shown above the page instead of "Chapter 2".
  final String title;
  final String titleTargetLang;
  /// The picture for this beat, chosen to match what happens here.
  final String emoji;
  final String narration;
  final String narrationTargetLang;
  final String interactionType; // choice | speak | listen
  final List<String> options;
  final String expectedIntent;
  /// Wanx-generated illustration URL (empty = use emoji fallback).
  final String illustrationUrl;

  /// Progressive generation: pending | generating | ready | failed.
  /// Classic stories are always 'ready'.
  final String status;

  ApprovedScene({
    required this.index,
    this.title = '',
    this.titleTargetLang = '',
    this.emoji = '',
    this.narration = '',
    this.narrationTargetLang = '',
    this.interactionType = 'listen',
    this.options = const [],
    this.expectedIntent = '',
    this.illustrationUrl = '',
    this.status = 'ready',
  });

  bool get isReady => status == 'ready';

  factory ApprovedScene.fromJson(Map<String, dynamic> json) {
    final interaction = json['interaction'] as Map<String, dynamic>? ?? {};
    return ApprovedScene(
      index: json['index'] as int? ?? 0,
      title: json['title'] as String? ?? '',
      titleTargetLang: json['title_target_lang'] as String? ?? '',
      emoji: json['emoji'] as String? ?? '',
      narration: json['narration'] as String? ?? '',
      narrationTargetLang: json['narration_target_lang'] as String? ?? '',
      interactionType: interaction['type'] as String? ?? 'listen',
      options: (interaction['options'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      expectedIntent: interaction['expected_intent'] as String? ?? '',
      illustrationUrl: json['illustration_url'] as String? ?? '',
      status: json['status'] as String? ?? 'ready',
    );
  }
}

/// A "word to learn" — Mina speaks it (vocab_N asset), the child repeats.
class VocabWord {
  final String word; // English
  final String wordTargetLang;
  final String romanised;

  VocabWord({
    required this.word,
    this.wordTargetLang = '',
    this.romanised = '',
  });

  factory VocabWord.fromJson(Map<String, dynamic> json) => VocabWord(
        word: json['word'] as String? ?? '',
        wordTargetLang: json['word_target_lang'] as String? ?? '',
        romanised: json['romanised'] as String? ?? '',
      );
}

/// F4 — the full approved story + media manifest for child mode.
class ApprovedStory {
  final String packageId;
  final String locale;
  final String title;
  final String titleTargetLang;
  final List<ApprovedScene> scenes;
  final String mission;
  final String missionTargetLang;
  final String handoffPrompt;
  final String handoffPromptTargetLang;
  final String handoffResponseSuggestion;
  final String familyVoiceMode; // F9 — confident_speaker | learning_parent
  final String speakingGoal;
  final String targetPhrase;
  final List<String> targetWords;
  final List<VocabWord> vocabulary;
  /// The chant that comes back every scene — the child joins in on it.
  final String refrain;
  final String refrainTargetLang;
  final List<MediaAsset> manifest;

  ApprovedStory({
    required this.packageId,
    this.locale = 'ta-SG',
    this.title = '',
    this.titleTargetLang = '',
    this.scenes = const [],
    this.mission = '',
    this.missionTargetLang = '',
    this.handoffPrompt = '',
    this.handoffPromptTargetLang = '',
    this.handoffResponseSuggestion = '',
    this.familyVoiceMode = 'confident_speaker',
    this.speakingGoal = '',
    this.targetPhrase = '',
    this.targetWords = const [],
    this.vocabulary = const [],
    this.refrain = '',
    this.refrainTargetLang = '',
    this.manifest = const [],
  });

  MediaAsset? assetFor(String id) {
    for (final a in manifest) {
      if (a.id == id) return a;
    }
    return null;
  }

  factory ApprovedStory.fromPackageJson(Map<String, dynamic> pkg) {
    final story = pkg['story'] as Map<String, dynamic>? ?? {};
    final mission = story['room_mission'] as Map<String, dynamic>? ?? {};
    final handoff = story['family_handoff'] as Map<String, dynamic>? ?? {};
    final media = pkg['media'] as Map<String, dynamic>? ?? {};
    final language = pkg['language'] as Map<String, dynamic>? ?? {};
    final familyVoice = pkg['family_voice'] as Map<String, dynamic>? ?? {};
    final plan = pkg['learning_plan'] as Map<String, dynamic>? ?? {};
    return ApprovedStory(
      packageId: pkg['id'] as String? ?? '',
      locale: language['locale'] as String? ?? 'ta-SG',
      title: story['title'] as String? ?? '',
      titleTargetLang: story['title_target_lang'] as String? ?? '',
      scenes: (story['scenes'] as List<dynamic>?)
              ?.map((e) => ApprovedScene.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      mission: mission['instruction'] as String? ?? '',
      missionTargetLang: mission['instruction_target_lang'] as String? ?? '',
      handoffPrompt: handoff['prompt'] as String? ?? '',
      handoffPromptTargetLang:
          handoff['prompt_target_lang'] as String? ?? '',
      handoffResponseSuggestion:
          handoff['response_suggestion'] as String? ?? '',
      familyVoiceMode: familyVoice['mode'] as String? ??
          handoff['mode'] as String? ??
          'confident_speaker',
      speakingGoal: plan['speaking_goal'] as String? ?? '',
      targetPhrase: plan['target_phrase'] as String? ?? '',
      targetWords: (plan['target_words'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      vocabulary: (story['vocabulary'] as List<dynamic>?)
              ?.map((e) => VocabWord.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      refrain: story['refrain'] as String? ?? '',
      refrainTargetLang: story['refrain_target_lang'] as String? ?? '',
      manifest: (media['manifest'] as List<dynamic>?)
              ?.map((e) => MediaAsset.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class SSEEvent {
  final String type;
  final String agent;
  final double progressPct;
  final String status;
  final String error;
  final String packageId;
  final String question; // F3 — needs_clarification
  // Book engine: how many pages are written so far, out of how many.
  final int cardsReady;
  final int cardsTotal;

  SSEEvent({
    required this.type,
    this.agent = '',
    this.progressPct = 0.0,
    this.status = '',
    this.error = '',
    this.packageId = '',
    this.question = '',
    this.cardsReady = 0,
    this.cardsTotal = 0,
  });

  factory SSEEvent.fromJson(Map<String, dynamic> json) => SSEEvent(
        type: json['type'] as String? ?? '',
        agent: json['agent'] as String? ?? '',
        progressPct: (json['progress_pct'] as num?)?.toDouble() ?? 0.0,
        status: json['status'] as String? ?? '',
        error: json['error'] as String? ?? '',
        packageId: json['package_id'] as String? ?? '',
        question: json['question'] as String? ?? '',
        cardsReady: (json['ready_cards'] as num?)?.toInt() ?? 0,
        cardsTotal: (json['total_cards'] as num?)?.toInt() ?? 0,
      );
}
